crates/common/src/paged_memory.rs

```diff
@@ -0,0 +1,341 @@
+use std::collections::HashMap;
+use std::ops::{Index, IndexMut};
+use stwo_prover::core::fields::qm31::QM31;
+
+// TODO sync this with VM crate
+pub const DEFAULT_MEMORY_SIZE: usize = 1 << 28;
+/// The default page size is 64K elements.
+pub const DEFAULT_PAGE_SIZE: usize = 1 << 16;
+
+/// Sparse, paged memory optimized for large logical address spaces.
+///
+/// - The logical memory is split into fixed-size pages. Each page is a contiguous
+///   native heap allocation (`Box<[QM31]>`).
+/// - Pages are allocated lazily on first write. Reads never allocate and return
+///   `None` for missing pages/cells; callers can treat that as zero.
+/// - `len` tracks the logical length (highest written index + 1, or the latest
+///   resize). It does not imply physical allocation for all covered pages.
+/// - This design keeps zero-filled regions sparse while providing fast, cache-friendly
+///   access within allocated pages.
+#[derive(Debug, Clone)]
+pub struct PagedMemory {
+    page_size: usize,
+    num_pages: usize,
+    len: usize,
+    pages: Vec<Option<Page>>, // index = page_number; lazy-allocated pages
```

Comment:
this would not help with the fact that you'll need to resize your vec for un-allocated page between the last page used and next used. We could consider a BTreeMap for ordered iteration